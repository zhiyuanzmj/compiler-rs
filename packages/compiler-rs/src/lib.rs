pub mod ir;
pub mod transform;
pub mod utils;
