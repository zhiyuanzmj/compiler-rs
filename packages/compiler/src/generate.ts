import { extend, remove } from '@vue/shared'
import { genBlockContent } from './generators/block'
import { genTemplates } from './generators/template'
import { setTemplateRefIdent } from './generators/templateRef'
import {
  buildCodeFragment,
  codeFragmentToString,
  INDENT_END,
  INDENT_START,
  NEWLINE,
} from './utils/generate'
import type { BlockIRNode, RootIRNode } from './ir'
import type { SimpleExpressionNode } from './utils'
import type { RawSourceMap } from 'source-map-js'

export type CodegenOptions = {
  /**
   * Generate source map?
   * @default false
   */
  sourceMap?: boolean
  /**
   * Filename for source map generation.
   * Also used for self-recursive reference in templates
   * @default 'index.jsx'
   */
  filename?: string
  templates?: string[]
}

export class CodegenContext {
  options: Required<CodegenOptions>

  helpers: Set<string> = new Set<string>([])

  helper = (name: string) => {
    this.helpers.add(name)
    return `_${name}`
  }

  delegates: Set<string> = new Set<string>()

  identifiers: Record<string, (string | SimpleExpressionNode)[]> =
    Object.create(null)

  block: BlockIRNode
  withId<T>(
    fn: () => T,
    map: Record<string, string | SimpleExpressionNode | null>,
  ): T {
    const { identifiers } = this
    const ids = Object.keys(map)

    for (const id of ids) {
      identifiers[id] ||= []
      identifiers[id].unshift(map[id] || id)
    }

    const ret = fn()
    ids.forEach((id) => remove(identifiers[id], map[id] || id))

    return ret
  }

  enterBlock(block: BlockIRNode) {
    const parent = this.block
    this.block = block
    return (): BlockIRNode => (this.block = parent)
  }

  scopeLevel: number = 0
  enterScope(): [level: number, exit: () => number] {
    return [this.scopeLevel++, () => this.scopeLevel--] as const
  }

  constructor(
    public ir: RootIRNode,
    options: CodegenOptions,
  ) {
    const defaultOptions: Required<CodegenOptions> = {
      sourceMap: false,
      filename: `index.jsx`,
      templates: [],
    }
    this.options = extend(defaultOptions, options)
    this.block = ir.block
  }
}

export interface VaporCodegenResult {
  ast: RootIRNode
  helpers: Set<string>
  templates: string[]
  delegates: Set<string>
  code: string
  map?: RawSourceMap
}

// IR -> JS codegen
export function generate(
  ir: RootIRNode,
  options: CodegenOptions = {},
): VaporCodegenResult {
  const [frag, push] = buildCodeFragment()
  const context = new CodegenContext(ir, options)
  const { helpers } = context

  push(INDENT_START)
  if (ir.hasTemplateRef) {
    push(
      NEWLINE,
      `const ${setTemplateRefIdent} = ${context.helper('createTemplateRefSetter')}()`,
    )
  }
  push(...genBlockContent(ir.block, context, true))
  push(INDENT_END, NEWLINE)

  if (context.delegates.size) {
    context.helper('delegateEvents')
  }
  const templates = genTemplates(ir.templates, ir.rootTemplateIndex, context)

  const [code, map] = codeFragmentToString(frag, context)

  return {
    code,
    ast: ir,
    map: map && map.toJSON(),
    helpers,
    templates,
    delegates: context.delegates,
  }
}
