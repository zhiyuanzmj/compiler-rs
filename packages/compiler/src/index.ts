export { compile, type CompilerOptions, type TransformPreset } from './compile'
export * from './transform'
export {
  CodegenContext,
  generate,
  type CodegenOptions,
  type VaporCodegenResult,
} from './generate'

export * from './ir'

export { transformText } from './transforms/transformText'
export { transformElement } from './transforms/transformElement'
export { transformChildren } from './transforms/transformChildren'
export { transformTemplateRef } from './transforms/transformTemplateRef'
export { transformVBind } from './transforms/vBind'
export { transformVOn } from './transforms/vOn'
export { transformVSlot } from './transforms/vSlot'
export { transformVSlots } from './transforms/vSlots'
export { transformVModel } from './transforms/vModel'
export { transformVShow } from './transforms/vShow'
export { transformVHtml } from './transforms/vHtml'
export { transformVFor } from './transforms/vFor'
export { transformVIf } from './transforms/vIf'
export { transformVOnce } from './transforms/vOnce'
export { transformVText } from './transforms/vText'
