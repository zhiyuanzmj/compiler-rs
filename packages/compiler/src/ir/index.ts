export {
  CreateNodesIRNode,
  _DynamicFlag as DynamicFlag,
  IRNode,
  IRNodeTypes,
  isBlockOperation,
  type BaseIRNode,
  type BlockIRNode,
  type CreateComponentIRNode,
  type DeclareOldRefIRNode,
  type DirectiveIRNode,
  type DirectiveNode,
  type ForIRNode,
  type GetTextChildIRNode,
  type IfIRNode,
  type InsertionStateTypes,
  type InsertNodeIRNode,
  type IRDynamicInfo,
  type IREffect,
  type IRFor,
  type OperationNode,
  type PrependNodeIRNode,
  type RootIRNode,
  type SetDynamicEventsIRNode,
  type SetDynamicPropsIRNode,
  type SetEventIRNode,
  type SetHtmlIRNode,
  type SetNodesIRNode,
  type SetPropIRNode,
  type SetTemplateRefIRNode,
  type SetTextIRNode,
  type SimpleExpressionNode,
  type SourceLocation,
} from '@vue-jsx-vapor/compiler-rs'

export * from './component'
