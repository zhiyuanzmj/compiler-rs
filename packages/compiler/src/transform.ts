import { extend, isArray, isString, NOOP } from '@vue/shared'
import {
  DynamicFlag,
  IRNodeTypes,
  type BlockIRNode,
  type IRDynamicInfo,
  type IRSlots,
  type OperationNode,
  type RootIRNode,
  type RootNode,
  type SetEventIRNode,
} from './ir'
import {
  findProp,
  getText,
  isConstantExpression,
  isConstantNode,
  isTemplate,
  newBlock,
  newDynamic,
  type CompilerError,
  type SimpleExpressionNode,
} from './utils'
import type { CodegenOptions } from './generate'
import type { JSXAttribute, JSXElement, JSXFragment } from 'oxc-parser'

export type NodeTransform = (
  node: BlockIRNode['node'],
  context: TransformContext<BlockIRNode['node']>,
) => void | (() => void) | (() => void)[]

export type DirectiveTransform = (
  dir: JSXAttribute,
  node: JSXElement,
  context: TransformContext<JSXElement>,
) => DirectiveTransformResult | void

export interface DirectiveTransformResult {
  key: SimpleExpressionNode
  value: SimpleExpressionNode
  modifier?: '.' | '^'
  runtimeCamelize?: boolean
  handler?: boolean
  handlerModifiers?: SetEventIRNode['modifiers']
  model?: boolean
  modelModifiers?: string[]
}

// A structural directive transform is technically also a NodeTransform;
// Only v-if and v-for fall into this category.
export type StructuralDirectiveTransform = (
  node: JSXElement,
  dir: JSXAttribute,
  context: TransformContext,
) => void | (() => void)

export type TransformOptions = CodegenOptions & {
  source?: string
  /**
   * Whether to compile components to createComponentWithFallback.
   * @default false
   */
  withFallback?: boolean
  /**
   * Indicates that transforms and codegen should try to output valid TS code
   */
  isTS?: boolean
  /**
   * An array of node transforms to be applied to every AST node.
   */
  nodeTransforms?: NodeTransform[]
  /**
   * An object of { name: transform } to be applied to every directive attribute
   * node found on element nodes.
   */
  directiveTransforms?: Record<string, DirectiveTransform | undefined>
  /**
   * Separate option for end users to extend the native elements list
   */
  isCustomElement?: (tag: string) => boolean | void
  onError?: (error: CompilerError) => void
}
const defaultOptions: Required<TransformOptions> = {
  source: '',
  sourceMap: false,
  filename: 'index.jsx',
  nodeTransforms: [],
  directiveTransforms: {},
  templates: [],
  isCustomElement: NOOP,
  isTS: false,
  withFallback: false,
  onError: (error: CompilerError) => {
    throw error
  },
}

export class TransformContext<
  T extends BlockIRNode['node'] = BlockIRNode['node'],
> {
  parent: TransformContext<RootNode | JSXElement | JSXFragment> | null = null
  root: TransformContext<RootNode>
  index: number = 0

  block: BlockIRNode
  options: Required<TransformOptions>

  template: string = ''
  childrenTemplate: (string | null)[] = []
  dynamic: IRDynamicInfo

  inVOnce: boolean = false
  inVFor: number = 0

  component: Set<string>
  directive: Set<string>

  slots: IRSlots[] = []

  private globalId = 0

  constructor(
    public ir: RootIRNode,
    public node: T,
    options: TransformOptions = {},
  ) {
    this.options = extend({}, defaultOptions, options)
    this.block = this.ir.block
    this.dynamic = this.ir.block.dynamic
    this.component = this.ir.component
    this.directive = this.ir.directive
    this.root = this as TransformContext<RootNode>
  }

  enterBlock(ir: BlockIRNode, isVFor: boolean = false): () => void {
    const { block, template, dynamic, childrenTemplate, slots } = this
    this.block = ir
    this.dynamic = ir.dynamic
    this.template = ''
    this.childrenTemplate = []
    this.slots = []
    isVFor && this.inVFor++
    return () => {
      // exit
      this.registerTemplate()
      this.block = block
      this.template = template
      this.dynamic = dynamic
      this.childrenTemplate = childrenTemplate
      this.slots = slots
      isVFor && this.inVFor--
    }
  }

  increaseId = () => this.globalId++
  reference() {
    if (this.dynamic.id !== undefined) return this.dynamic.id
    this.dynamic.flags |= DynamicFlag.REFERENCED
    return (this.dynamic.id = this.increaseId())
  }

  pushTemplate(content: string) {
    const existing = this.ir.templates.indexOf(content)
    if (existing !== -1) return existing
    this.ir.templates.push(content)
    return this.ir.templates.length - 1
  }

  registerTemplate() {
    if (!this.template) return -1
    const id = this.pushTemplate(this.template)
    return (this.dynamic.template = id)
  }

  registerEffect(
    expressions: SimpleExpressionNode[],
    operation: OperationNode | OperationNode[],
    getEffectIndex = (): number => this.block.effect.length,
    getOperationIndex = (): number => this.block.operation.length,
  ) {
    const operations = [operation].flat()
    expressions = expressions.filter((exp) => !isConstantExpression(exp))
    if (
      this.inVOnce ||
      expressions.length === 0 ||
      expressions.every((e) => e.ast && isConstantNode(e.ast))
    ) {
      return this.registerOperation(operations, getOperationIndex)
    }

    this.block.effect.splice(getEffectIndex(), 0, {
      expressions,
      operations,
    })
  }

  registerOperation(
    operation: OperationNode | OperationNode[],
    getOperationIndex = (): number => this.block.operation.length,
  ) {
    this.block.operation.splice(getOperationIndex(), 0, ...[operation].flat())
  }

  create<E extends T>(node: E, index: number): TransformContext<E> {
    return Object.assign(Object.create(TransformContext.prototype), this, {
      node,
      parent: this as any,
      index,

      template: '',
      childrenTemplate: [],
      dynamic: newDynamic(),
    } satisfies Partial<TransformContext<T>>)
  }
}

// AST -> IR
export function transform(
  node: RootNode,
  options: TransformOptions = {},
): RootIRNode {
  const ir: RootIRNode = {
    type: IRNodeTypes.ROOT,
    node,
    source: node.source,
    templates: options.templates || [],
    component: new Set(),
    directive: new Set(),
    block: newBlock(node),
    hasTemplateRef: false,
  }

  const context = new TransformContext(ir, node, options)

  transformNode(context)

  return ir
}

export function transformNode(context: TransformContext<BlockIRNode['node']>) {
  let { node } = context

  // apply transform plugins
  const { nodeTransforms } = context.options
  const exitFns = []
  for (const nodeTransform of nodeTransforms) {
    const onExit = nodeTransform(node, context)
    if (onExit) {
      if (isArray(onExit)) {
        exitFns.push(...onExit)
      } else {
        exitFns.push(onExit)
      }
    }
    if (!context.node) {
      // node was removed
      return
    } else {
      // node may have been replaced
      node = context.node
    }
  }

  // exit transforms
  context.node = node
  let i = exitFns.length
  while (i--) {
    exitFns[i]()
  }

  if (context.node.type === IRNodeTypes.ROOT) {
    context.registerTemplate()
  }
}

export function createStructuralDirectiveTransform(
  name: string | string[],
  fn: StructuralDirectiveTransform,
): NodeTransform {
  const matches = (n: string) =>
    isString(name) ? n === name : name.includes(n)

  return (node, context) => {
    if (node.type === 'JSXElement') {
      const {
        openingElement: { attributes },
      } = node
      // structural directive transforms are not concerned with slots
      // as they are handled separately in vSlot.ts
      if (isTemplate(node) && findProp(node, 'v-slot')) {
        return
      }
      const exitFns = []
      for (const prop of attributes) {
        if (prop.type !== 'JSXAttribute') continue
        const propName = getText(prop.name, context)
        if (propName.startsWith('v-') && matches(propName.slice(2))) {
          attributes.splice(attributes.indexOf(prop), 1)
          const onExit = fn(node, prop, context as TransformContext)
          if (onExit) exitFns.push(onExit)
          break
        }
      }
      return exitFns
    }
  }
}
