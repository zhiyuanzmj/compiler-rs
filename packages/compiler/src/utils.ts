export * from './utils/error'
export * from './utils/expression'
export * from './utils/ast'
export * from './utils/directive'
export * from './utils/text'
export * from './utils/transform'
export * from './utils/generate'
export * from './utils/dom'
