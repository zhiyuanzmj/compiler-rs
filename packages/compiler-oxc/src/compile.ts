import { extend, isString } from '@vue/shared'
import { generate, type VaporCodegenResult } from './generate'
import { IRNodeTypes, type RootNode } from './ir'
import {
  transform,
  type DirectiveTransform,
  type NodeTransform,
  type TransformOptions,
} from './transform'
import { transformChildren } from './transforms/transformChildren'
import { transformElement } from './transforms/transformElement'
import { transformTemplateRef } from './transforms/transformTemplateRef'
import { transformText } from './transforms/transformText'
import { transformVBind } from './transforms/vBind'
import { transformVFor } from './transforms/vFor'
import { transformVHtml } from './transforms/vHtml'
import { transformVIf } from './transforms/vIf'
import { transformVModel } from './transforms/vModel'
import { transformVOn } from './transforms/vOn'
import { transformVOnce } from './transforms/vOnce'
import { transformVShow } from './transforms/vShow'
import { transformVSlot } from './transforms/vSlot'
import { transformVSlots } from './transforms/vSlots'
import { transformVText } from './transforms/vText'
import { parseExpression } from './utils'
import type { JSXElement, JSXFragment } from 'oxc-parser'

// code/AST -> IR (transform) -> JS (generate)
export function compile(
  source: JSXElement | JSXFragment | string,
  options: CompilerOptions = {},
): VaporCodegenResult {
  const resolvedOptions = extend({}, options, {
    filename: 'index.jsx',
  })
  if (!resolvedOptions.source && isString(source)) {
    resolvedOptions.source = source
  }
  const root = isString(source)
    ? parseExpression(resolvedOptions.filename, source)
    : source
  const children =
    root.type === 'JSXFragment'
      ? root.children
      : root.type === 'JSXElement'
        ? [root]
        : []
  const ast: RootNode = {
    type: IRNodeTypes.ROOT,
    children,
    source: resolvedOptions.source || '',
  }
  const [nodeTransforms, directiveTransforms] = getBaseTransformPreset()

  const ir = transform(
    ast,
    extend({}, resolvedOptions, {
      nodeTransforms: [
        ...nodeTransforms,
        ...(resolvedOptions.nodeTransforms || []), // user transforms
      ],
      directiveTransforms: extend(
        {},
        directiveTransforms,
        resolvedOptions.directiveTransforms || {}, // user transforms
      ),
    }),
  )

  return generate(ir, resolvedOptions)
}

export type CompilerOptions = TransformOptions
export type TransformPreset = [
  NodeTransform[],
  Record<string, DirectiveTransform>,
]

export function getBaseTransformPreset(): TransformPreset {
  return [
    [
      transformVOnce,
      transformVIf,
      transformVFor,
      transformTemplateRef,
      transformElement,
      transformText,
      transformVSlot,
      transformChildren,
    ],
    {
      bind: transformVBind,
      on: transformVOn,
      model: transformVModel,
      show: transformVShow,
      html: transformVHtml,
      text: transformVText,
      slots: transformVSlots,
    },
  ]
}
