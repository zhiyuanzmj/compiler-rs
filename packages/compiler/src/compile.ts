import { extend, isString } from '@vue/shared'
import { generate, type VaporCodegenResult } from './generate'
import { IRNodeTypes } from './ir'
import { transform, type TransformOptions } from './transform'
import { parseExpression } from './utils'
import type { JSXElement, JSXFragment } from 'oxc-parser'

// code/AST -> IR (transform) -> JS (generate)
export function compile(
  source: JSXElement | JSXFragment | string,
  options: CompilerOptions = {},
): VaporCodegenResult {
  const resolvedOptions = extend({}, options, {
    filename: 'index.jsx',
  })
  if (!resolvedOptions.source && isString(source)) {
    resolvedOptions.source = source
  }
  const root = isString(source)
    ? parseExpression(resolvedOptions.filename, source)
    : source
  const children =
    root.type === 'JSXFragment'
      ? root.children
      : root.type === 'JSXElement'
        ? [root]
        : []
  const ast = {
    type: IRNodeTypes.ROOT,
    children,
  } as unknown as Node

  const ir = transform(ast, extend({}, resolvedOptions))

  return generate(ir, resolvedOptions)
}

export type CompilerOptions = TransformOptions
