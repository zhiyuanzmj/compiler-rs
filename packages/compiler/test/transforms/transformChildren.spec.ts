import { describe, expect, test } from 'vitest'
import { IRNodeTypes } from '../../src'
import { makeCompile } from './_utils'

const compileWithElementTransform = makeCompile()

describe('compiler: children transform', () => {
  test('basic', () => {
    const { code, helpers } = compileWithElementTransform(
      `<div>
        {foo} {bar}
       </div>`,
    )
    expect(code).toMatchInlineSnapshot(`
      "
        const n0 = t0()
        const x0 = _child(n0)
        _setNodes(x0, () => (foo), " ", () => (bar))
        return n0
      "
    `)
    expect(helpers).contains.all.keys('setNodes')
  })

  test('comments', () => {
    const { code } = compileWithElementTransform(
      '<>{/*foo*/}<div>{/*bar*/}</div></>',
    )
    expect(code).toMatchInlineSnapshot(`
      "
        const n1 = t0()
        return n1
      "
    `)
  })

  test('fragment', () => {
    const { code } = compileWithElementTransform('<>{foo}</>')
    expect(code).toMatchInlineSnapshot(`
      "
        const n0 = _createNodes(() => (foo))
        return n0
      "
    `)
  })

  test('children & sibling references', () => {
    const { code, helpers } = compileWithElementTransform(
      `<div id={id}>
        <p>{ first }</p>
        123 { second } 456 {foo}
        <p>{ forth }</p>
      </div>`,
    )
    expect(code).toMatchInlineSnapshot(`
      "
        const n3 = t0()
        const n0 = _child(n3)
        const n1 = _next(n0)
        const n2 = _next(n1)
        const x0 = _child(n0)
        _setNodes(x0, () => (first))
        _setNodes(n1, "123 ", () => (second), " 456 ", () => (foo))
        const x2 = _child(n2)
        _setNodes(x2, () => (forth))
        _renderEffect(() => _setProp(n3, "id", id))
        return n3
      "
    `)
    expect(Array.from(helpers)).containSubset([
      'child',
      'renderEffect',
      'next',
      'setNodes',
      'template',
    ])
  })

  test('{...obj}', () => {
    const { code, ir } = compileWithElementTransform(`<div {...obj} />`)
    expect(code).toMatchInlineSnapshot(`
      "
        const n0 = t0()
        _renderEffect(() => _setDynamicProps(n0, [obj], true))
        return n0
      "
    `)
    expect(ir.block.effect).toMatchObject([
      {
        expressions: [],
        operations: [
          {
            type: IRNodeTypes.SET_DYNAMIC_PROPS,
            element: 0,
            props: [
              {
                value: {
                  content: 'obj',
                  isStatic: false,
                },
              },
            ],
          },
        ],
      },
    ])
    expect(code).contains('_setDynamicProps(n0, [obj], true)')
  })

  test('efficient traversal', () => {
    const { code } = compileWithElementTransform(
      `<div>
    <div>x</div>
    <div><span>{ msg }</span></div>
    <div><span>{ msg }</span></div>
    <div><span>{ msg }</span></div>
  </div>`,
    )
    expect(code).toMatchSnapshot()
  })

  test('efficient find', () => {
    const { code } = compileWithElementTransform(
      `<div>
        <div>x</div>
        <div>x</div>
        <div>{ msg }</div>
      </div>`,
    )
    expect(code).contains(`const n0 = _nthChild(n1, 2)`)
    expect(code).toMatchSnapshot()
  })

  test('anchor insertion in middle', () => {
    const { code, ir } = compileWithElementTransform(
      `<div>
        <div></div>
        <div v-if={1}></div>
        <div></div>
      </div>`,
    )
    // ensure the insertion anchor is generated before the insertion statement
    expect(code).toMatch(`const n3 = _next(_child(n4))`)
    expect(code).toMatch(`_setInsertionState(n4, n3)`)
    expect(code).toMatchSnapshot()
    expect(ir.templates).toMatchInlineSnapshot(`
      [
        "<div></div>",
        "<div><div></div><!><div></div></div>",
      ]
    `)
  })

  test('JSXComponent in JSXExpressionContainer', () => {
    const { code } = compileWithElementTransform(
      `<div>
        {<Comp />}
      </div>`,
    )
    expect(code).contains(`_setNodes(x0, () => (<Comp />))`)
    expect(code).toMatchSnapshot()
  })
})
