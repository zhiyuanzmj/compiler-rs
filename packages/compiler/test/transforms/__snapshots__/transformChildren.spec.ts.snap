// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiler: children transform > JSXComponent in JSXExpressionContainer 1`] = `
"
  const n0 = t0()
  const x0 = _child(n0)
  _setNodes(x0, () => (<Comp />))
  return n0
"
`;

exports[`compiler: children transform > anchor insertion in middle 1`] = `
"
  const n4 = t1()
  const n3 = _next(_child(n4))
  _setInsertionState(n4, n3)
  const n0 = _createIf(() => (1), () => {
    const n2 = t0()
    return n2
  }, null, true)
  return n4
"
`;

exports[`compiler: children transform > efficient find 1`] = `
"
  const n1 = t0()
  const n0 = _nthChild(n1, 2)
  const x0 = _child(n0)
  _setNodes(x0, () => (msg))
  return n1
"
`;

exports[`compiler: children transform > efficient traversal 1`] = `
"
  const n3 = t0()
  const p0 = _next(_child(n3))
  const n0 = _child(p0)
  const p1 = _next(p0)
  const n1 = _child(p1)
  const p2 = _next(p1)
  const n2 = _child(p2)
  const x0 = _child(n0)
  _setNodes(x0, () => (msg))
  const x1 = _child(n1)
  _setNodes(x1, () => (msg))
  const x2 = _child(n2)
  _setNodes(x2, () => (msg))
  return n3
"
`;
